teps/0090-matrix.md

```diff
@@ -596,17 +627,546 @@ For example:
+the `Matrix` have been executed.
+
+There are two types of dependencies between `PipelineTasks`:
+* Ordering: declared using `runAfter`.
+* Resource: established by passing resources, such as `Results`.
+
+#### Ordering Dependencies
+
+In the example below, the *test* `PipelineTask` is fanned out using the *shards* 
+`Result` in the `Matrix`. The *build* `PipelineTask` is ordering dependent on 
+the *test* `PipelineTask`, based on `runAfter`. As such, all the *test* `TaskRuns` 
+have to complete execution successfully before *build* `PipelineTask` is executed.
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: pipeline
+spec:
+  tasks:
+  - name: clone
+    taskRef:
+      name: git-clone
+    ...
+  - name: get-test-shards
+    taskRef:
+      name: get-test-shards
+    runAfter:
+    - clone
+    ...
+  - name: test
+    taskRef:
+      name: test
+    matrix:
+    - name: shards
+      value: $(tasks.get-test-shards.results.shards)
+    ...
+  - name: build
+    taskRef:
+      name: build
+    runAfter:
+    - test
+    ...
+```
+
+#### Resource Dependencies
+
+In the example below, the *test* `PipelineTask` is fanned out using the *browsers*
+`Parameter` in the `Matrix`. The *notify* `PipelineTask` is resource dependent on
+the *test* `PipelineTask`, based on *output* `Results`. As such, all the *test* 
+`TaskRuns` have to complete execution successfully before *notify* `PipelineTask` 
+is executed.
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: pipeline
+spec:
+  params:
+  - name: browsers
+    type: array
+    default:
+      - chrome
+      - firefox
+      - safari
+  tasks:
+  - name: clone
+    taskRef:
+      name: git-clone
+    ...
+  - name: test
+    taskRef:
+      name: test
+    matrix:
+    - name: browsers
+      value: $(params.browsers)
+    ...
+  - name: notify
+    taskRef:
+      name: notify
+    params:
+    - name: messages
+      value: 
+      - $(tasks.test-1.results.output)
+      - $(tasks.test-2.results.output)
+      - $(tasks.test-3.results.output)
+    ...
+```
+
+## Implementation and Promotion Plan
+
+Access to the `matrix` feature and field will be guarded by the `alpha` feature gate. 
+This will give us a chance to gather feedback from users and iterate on the design 
+before promoting it to `beta`. 
+
+In addition, the feature will be implemented in a phases to ensure we handle the 
+complexities carefully. 
+
+#### Phase 1: API Change and Validation
+* Add the `matrix` API field.
+* Guard the `matrix` behind the `alpha` feature gate.
+* Implement validation of the `matrix` field, including:
+  * `Parameter` type must be `Array`.
+  * `Parameter` can be in only one of `matrix` or `params`, not both.
+  * `Parameter` must not have `Result` references.
+  * `Parameter` must substitute valid variables, such as `Execution Status`. 
+
+#### Phase 2: Execute TaskRuns
+* Implement fanning out of `PipelineTasks` into `TaskRuns` based on the `matrix`.
```

Comment:
Separating phase1 and phase2 will reduce the size of the PR, but I would prefer to at least keep them together in a release.

Reply:
Workspaces and all other `PipelineTask` features would be supported here, except for results and `Runs`, right?

Reply:
good idea, collapsed phase 1 and phase 2 into one milestone to be included in one release (but still in separate pull request)

Reply:
that's right, all pipelinetask features would be supported except results and runs (in milestone 1)